File: maximal-square/grandyang1.java
//grandyang idea1, 59ms, faster than 4%, not understood
class Solution {
    public int maximalSquare(char[][] matrix) {
        int res = 0;
        int m = matrix.length;
        
        
        for (int i = 0; i < m; ++i) {
            int n = matrix[i].length;
            int [] v = new int[n];
            for (int j = i; j < m; ++j) {
                for (int k = 0; k < matrix[j].length; ++k) {
                    if (matrix[j][k] == '1') ++v[k];
                }
                res = Math.max(res, getSquareArea(v, j - i + 1));
            }
        }
        return res;
    }
    private int getSquareArea(int []v, int k) {
        if (k > v.length) return 0;
        int count = 0;
        for (int i = 0; i < v.length; ++i) {
            if (v[i] != k) count = 0; 
            else ++count;
            if (count == k) return k * k;
        }
        return 0;
    }
}

File: count-of-range-sum/grandyang1.cpp
/*
grandyang idea1
Runtime: 40 ms, faster than 89.36% of C++ online submissions for Count of Range Sum.
Memory Usage: 12.3 MB, less than 66.67% of C++ online submissions for Count of Range Sum.

*/
class Solution {
public:
    int countRangeSum(vector<int>& nums, int lower, int upper) {
        int res = 0;
        long long sum = 0;
        multiset<long long> sums;
        sums.insert(0);
        for (int i = 0; i < nums.size(); ++i) {
            sum += nums[i];
            res += distance(sums.lower_bound(sum - upper), sums.upper_bound(sum - lower));
            sums.insert(sum);
        }
        return res;
    }        
};
